### PR TITLE
Version history, InfoMessage Log and notification log

### PR DESCRIPTION
**Pull request by zwim**
This fixes ancient  https://github.com/koreader/koreader/issues/1257

On every start of KOReader it is checked if the current git revision matches the revision of the last KOReader start (and if the device model of the last start matches the current model).

The git-rev and device model (and the date of the start) is stored in the file `version.log` if a change in the revision or the model (but not the date of course) is detected. This might simplify bug/regression detection.

The proposed solution covers OTA as well as manual updates.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10128)

### CODE REVIEW

File: frontend/version.lua [hunk 1]
```diff
@@ -86,4 +89,54 @@ function Version:getBuildDate()
     return self.date
 end
 
+-- Returns the last line in a file. If `skip_nl_at_end` is true, omit the NL at eof
```
Comment (NiLuJe) [inline frontend/version.lua:92]: You can just use `read("*line")` instead of doing LN gymnastics?

File: frontend/version.lua [hunk 2]
```diff
@@ -86,4 +89,54 @@ function Version:getBuildDate()
...
+            end
+        end
+    end
+
+    local lastLine = file:read("*a")
+    file:close()
+    return lastLine
+end
+
+--- Returns the KOReader git-rev and model of the last line in the VERSION\_LOG\_FILE
```
Comment (NiLuJe) [inline frontend/version.lua:116]: nit: stray escaping

File: frontend/version.lua [hunk 3]
```diff
@@ -86,4 +89,54 @@ function Version:getBuildDate()
...
+    local last_version_line = readLastLine(VERSION_LOG_FILE) or ""
+    self.last_version = last_version_line:match("%S*") or LAST_VERSION_NOT_FOUND
+    local model = last_version_line:match(" .* %(") or "  (" -- two spaces and a parens
+    model = model:sub(2, model:len() - 2)
+
+    return self.last_version, model
+end
+
+--- Appends KOReader git-rev, model and current date to the VERSION\_LOG\_FILE
+--- in the format 'git-rev "model" (YYYY-mm-dd HH:MM:SS)'
```
Comment (NiLuJe) [inline frontend/version.lua:128]: I'd probably simplify this to makie it more easily machine readable (i.e., CSV), then you can just use a gmatch loop and be done with it.

**Comment by zwim** [comment]
Also fixes #3418.

Added the module `logfile.lua` which is used by `reader.lua`, `infomessage.lua` and `notification.lua`.
The new module does not depend on any other KOReader module, so it can be easily included everywhere.

The number of lines on each log-file is/can be restricted so that only the last N lines will be kept. The number of log-lines is cached, so that file accessed are reduced. 
(Improvements could be to cache the whole log file content, but I am not sure, if I want to waste that memory;
or to reduce the line number to half, if the limit is reached. Don't know what is smarter, yet.)

The new module should be halfway future proof, as it can handle numerous log files.

Additionally, a toggle switch in the UI to enable/disable the feature will be implemented (by means of a hold-press).

Not thoroughly tested, as it is still draft.


The UI might look like this (the hold-press) is not implemented yet.


grafik

The Notification log:
grafik

The info-message log:
grafik


As always: Improvements are appreciated.

**Comment by poire-z** [comment]
I don't care about verstion history log, but I absolutely don't want any useless info & notification log ?!! Why ?! "Opening foobar.epub...", "Contrast set to 2" ?!

**Comment by zwim** [comment]
Maybe you ask the author of #3418 why he wants that ;-)

Anyway, if this PR is not wanted, I can and will close it. (Before wasting any more time on it.)

**Comment by NiLuJe** [comment]
While I sorta share the sentiment (notification history is *very* useful on Android, for instance. Here... not so much? ;o)), the impact on the affected module seems fairly minimal?

**Comment by NiLuJe** [comment]
> Maybe you ask the author of #3418 why he wants that ;-)

The age of that FR seems to be a strong hint to me that *nobody* really wants that? ^^

**Comment by zwim** [comment]
Fine by me.

**Comment by poire-z** [comment]
For the FR, a "Show last notifications" (stored in memory) could be enough - but I stlll find it not useful.
(Also, not all FR, because they are there, are worth being implemented :)

What's stupid with loging notification and InfoMessage is that you'll be logging tons of help_text and stuff, things possibly relevant and things totally not relevant.
If you ever wanted to log book openings and settings change (x), you would log _that_, the opening and the setting change, and not the content of all their common vehicles (Notification and InfoMessage). (And what about ConfirmBox, and the cancel|ok chosen. And if a Notifcation "Book style tweak saved" is useful, then, I request all my menu opening and navigation and selection to be logged too :)

(x): which I could understand the usefulness to some, knowing when you changed things, but I still find not worth adding - or we could just log that in the crash.log, we already do for book openings, that I sometimes find useful to remember when I opened that.

> I don't care about verstion history log

I meant: I don't care = I don't mind, not useful to me, but could make sense to some (unlike Nofication & Infomessage logs), so you don't have to close all this PR :/

**Comment by zwim** [comment]
@poire-z  Thanks for the clarification. But as English is a foreign language to me:
Do you mean I should keep the part of the `version.log` and drop the rest (infomessage and notifications)?

**Comment by zwim** [comment]
@poire-z  Don't worry, I won't start closing anything, which is not mine.

**Comment by Frenzie** [comment]
> Maybe you ask the author of #3418 why he wants that ;-)

Haha, well, it looks to me like I read a blogpost and said to myself that it'd be cool to have. Your example shows that perhaps it doesn't hold up to closer inspection, but perhaps it merely shows that some notifications are more important than others.

**Comment by poire-z** [comment]
> Do you mean I should keep the part of the version.log and drop the rest (infomessage and notifications)?

I mean : I can't say you _should_ :), but I'd say you _may_ keep the part of the version.log (because even if I don't personally see the need, it might be of help to people reporting issues and helping bissecting).

And yes, please drop the rest (infomessage and notifications).
(Or if you really want to keep some of the fun and address @Frenzie one time idea, keep some notification text in memory and add a single menu item "Show last notifications" in your Notifications submenu to show the last 10 ones - I don't think it deserves to be more featureful and complicated than that.)

**Comment by Frenzie** [comment]
> keep some notification text in memory and add a single menu item "Show last notifications" in your Notifications submenu to show the last 10 ones - I don't think it deserves to be more featureful and complicated than that.

Something that simple is what I had in mind, but the fact that I never actually wrote it is probably telling. ;-)

**Comment by zwim** [comment]
After sleeping/thinking about the whole thing, I think it is not worth the potential trouble of modifying `reader.lua` (and introduce a new functionality).
So I will close this.

**Comment by Frenzie** [comment]
If that's how you feel we should also close the issue? I don't  have any trouble with adding a couple of lines to reader.lua though. :-)

**Pull request closed without merging**